src/v/cloud_storage/remote.cc

```diff
@@ -106,7 +107,13 @@ static error_outcome categorize_error(
             code.value() != ECONNREFUSED && code.value() != ENETUNREACH
             && code.value() != ETIMEDOUT && code.value() != ECONNRESET
             && code.value() != EPIPE) {
-            vlog(ctxlog.error, "System error {}", cerr);
+            // When a bad file descriptor is seen and the leadership is lost,
+            // downgrade to a warning because gc could have removed the segment
+            if (code.value() == EBADF && !is_leader) {
```

Comment:
TBH the leadership check looks quite out of place at this layer of abstraction - I like the read lock solution better. We can then add the predicate (generically named, something like `should_continue`) as an optimization.

Reply:
The issues I saw with read-lock is that it
* makes us upload a file which is no longer required
* holds up the gc code flow while the lock is held

On an overall scheme of things, I think getting the system error is good because it stops us from uploading the file which we do not need and allows segment cleanup to keep working. It just needs to be managed and reported correctly.

I agree that the leadership check looks clunky, I tried to avoid it initially, but the issue is that we want to avoid the error log in the `categorize_error` and to avoid that we have to propagate the state to that function. 

I also thought to move out the logging from `categorize_error` and let the caller decide on the logging aspect and just let `categorize_error` do the classification. 

I think that is closer to @rystsov's suggestion, for this `categorize_error` will need to return some more context, it will require changes to all callers of `upload_segment` and other similar functions, but I can explore that. Will try to make this change.

Reply:
changed leadership check to generic struct `should_continue`, also moved the code out of `categorize_error`, logging is done in `upload_segment` based on predicate and exception to decide if it is warn/error.